pkg/ddl/notifier/events.go

```diff
@@ -370,55 +385,46 @@ func (s *SchemaChangeEvent) GetRemovePartitioningInfo() (
 	newSingleTableInfo *model.TableInfo,
 	droppedPartInfo *model.PartitionInfo,
 ) {
-	intest.Assert(s.tp == model.ActionRemovePartitioning)
-	return s.oldTableID4Partition, s.tableInfo, s.droppedPartInfo
+	intest.Assert(s.inner.Tp == model.ActionRemovePartitioning)
+	return s.inner.OldTableID4Partition, s.inner.TableInfo, s.inner.DroppedPartInfo
 }
 
 // NewFlashbackClusterEvent creates a schema change event whose type is
 // ActionFlashbackCluster.
 func NewFlashbackClusterEvent() *SchemaChangeEvent {
 	return &SchemaChangeEvent{
-		tp: model.ActionFlashbackCluster,
+		inner: &jsonSchemaChangeEvent{
+			Tp: model.ActionFlashbackCluster,
+		},
 	}
 }
 
+// jsonSchemaChangeEvent is used by SchemaChangeEvent when needed to (un)marshal data,
+// we want to hide the details to subscribers, so SchemaChangeEvent contain this struct.
 type jsonSchemaChangeEvent struct {
-	TableInfo            *model.TableInfo
-	OldTableInfo         *model.TableInfo
-	AddedPartInfo        *model.PartitionInfo
-	DroppedPartInfo      *model.PartitionInfo
-	ColumnInfos          []*model.ColumnInfo
+	TableInfo       *model.TableInfo
```

Comment:
add omitempty in json tag

Reply:
updated in https://github.com/pingcap/tidb/pull/56089/commits/dbd3445e1294feed31627a874061ec8a15b79029 and https://github.com/pingcap/tidb/pull/56089/commits/fe72a490bdff4ed6aa5e335091573416af659b93